 cutils: realtime scheduler class support

In order for Android processes to really take advantage of Linux's
real-time scheduler, set_sched_policy() needs to expose this
capability.  Real-time priorities are set via the sched_setscheduler()
API rather than setpriority().  We currently schedule them as benignly
as possible (using priority 1 and round-robin scheduling).

diff --git a/include/cutils/sched_policy.h b/include/cutils/sched_policy.h
index b01f6f7..ac58190 100644
--- a/include/cutils/sched_policy.h
+++ b/include/cutils/sched_policy.h
@@ -30,6 +30,7 @@ typedef enum {
     SP_AUDIO_APP  = 3,
     SP_AUDIO_SYS  = 4,
     SP_TOP_APP    = 5,
+    SP_REALTIME   = 6, /* Motorola, rknize2, 05/10/2013, IKJBXLINE-9555 */
     SP_CNT,
     SP_MAX        = SP_CNT - 1,
     SP_SYSTEM_DEFAULT = SP_FOREGROUND,
diff --git a/libcutils/sched_policy.c b/libcutils/sched_policy.c
index b01f6f7..ac58191 100644
--- a/libcutils/sched_policy.c
+++ b/libcutils/sched_policy.c
@@ -259,6 +259,10 @@ int get_sched_policy(int tid, SchedPolicy *policy)
             *policy = SP_FOREGROUND;
         else if (rc == SCHED_BATCH)
             *policy = SP_BACKGROUND;
+/* BEGIN Motorola, rknize2, 05/10/2013, IKJBXLINE-9555 */
+        else if (rc == SCHED_RR)
+            *policy = SP_REALTIME;
+/* END Motorola, IKJBXLINE-9555 */
         else {
             errno = ERANGE;
             return -1;
@@ -365,13 +369,22 @@ int set_sched_policy(int tid, SchedPolicy policy)
     case SP_SYSTEM:
         SLOGD("/// tid %d (%s)", tid, thread_name);
         break;
+/* BEGIN Motorola, rknize2, 05/10/2013, IKJBXLINE-9555 */
+    case SP_REALTIME:
+        SLOGD("!!! tid %d (%s)", tid, thread_name);
+        break;
+/* END Motorola, IKJBXLINE-9555 */
     default:
         SLOGD("??? tid %d (%s)", tid, thread_name);
         break;
     }
 #endif
 
-    if (__sys_supports_schedgroups) {
+/* BEGIN Motorola, rknize2, 05/10/2013, IKJBXLINE-9555
+ * Schedule groups are not supported for RT processes. */
+    if (__sys_supports_schedgroups &&
+        policy != SP_REALTIME) {
+/* END Motorola, IKJBXLINE-9555 */
         int boost_fd = -1;
         switch (policy) {
         case SP_BACKGROUND:
@@ -398,12 +411,21 @@ int set_sched_policy(int tid, SchedPolicy policy)
 #endif
     } else {
         struct sched_param param;
+/* BEGIN Motorola, rknize2, 05/10/2013, IKJBXLINE-9555
+ * Allow the RT policy at the lowest priority. */
+        int posix_policy = SCHED_NORMAL;
+
+        param.sched_priority = 0; /* unused for non-RT policies */
+        if (policy == SP_BACKGROUND) {
+            posix_policy = SCHED_BATCH;
+        } else if (policy == SP_REALTIME) {
+            posix_policy = SCHED_RR;
+            param.sched_priority = 1; /* lowest RT priority */
+        }
 
-        param.sched_priority = 0;
-        sched_setscheduler(tid,
-                           (policy == SP_BACKGROUND) ?
-                           SCHED_BATCH : SCHED_NORMAL,
-                           &param);
+        if (sched_setscheduler(tid, posix_policy, &param) < 0)
+            SLOGE("sched_setscheduler failed: tid %d, errno=%d", tid, errno);
+/* END Motorola, IKJBXLINE-9555 */
     }
 
     prctl(PR_SET_TIMERSLACK_PID,
@@ -439,6 +461,7 @@ const char *get_sched_policy_name(SchedPolicy policy)
        [SP_AUDIO_APP]  = "aa",
        [SP_AUDIO_SYS]  = "as",
        [SP_TOP_APP]    = "ta",
+       [SP_REALTIME]   = "rt", /* Motorola, w04904, 05/10/2013, IKJBXLINE-9555 */
     };
     if ((policy < SP_CNT) && (strings[policy] != NULL))
         return strings[policy];
